### PR TITLE
Store creator hash prefixes in a separate lookup file

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/43994

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Regression testing:

### Clean Profile Case

- Start the browser.
- Enable Rewards.
- Visit a verified publisher (e.g. `laurenwags.github.io` on staging).
- Verify that the Rewards icon shows a "verified" badge.

### Upgrade Case

- With an existing profile with Rewards enabled, start the browser.
- Visit a verified publisher.
- Verify that the Rewards icon shows a "verified" badge.

### CODE REVIEW

File: components/brave_rewards/core/engine/hash_prefix_iterator.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+#define BRAVE_COMPONENTS_BRAVE_REWARDS_CORE_ENGINE_HASH_PREFIX_ITERATOR_H_
+
+#include <iterator>
+#include <string_view>
+
+namespace brave_rewards {
+
+// A random-access iterator over prefixes stored in an uncompressed prefix list,
+// suitable for binary search.
+class HashPrefixIterator {
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_iterator.h:16]: Hm, do we really need this thing?
If it's only to support std::binary_search, we can just compare [0.. count-1] indexes + a custom compare function instead.

File: components/brave_rewards/core/engine/hash_prefix_store.cc [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+std::optional<FileParts> ParseFile(base::span<const uint8_t> bytes) {
+  FileParts parts;
+  base::SpanReader reader(bytes);
+  if (!reader.ReadU32LittleEndian(parts.version)) {
+    return std::nullopt;
+  }
+  if (!reader.ReadU32LittleEndian(parts.prefix_size)) {
+    return std::nullopt;
+  }
+  parts.prefixes = reader.remaining_span();
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_store.cc:55]: I believe we need to verify size here. Otherwise we can get an overflow in case the file is broken (i.e. with a sudden shutdown)

File: ios/browser/api/brave_rewards/brave_rewards_api.mm [hunk 3]
```diff
@@ -42,6 +44,26 @@
...
+    : public RemoteWorker<brave_rewards::mojom::HashPrefixStore> {
+ public:
+  HashPrefixStoreWorker() : RemoteWorker(CreateTaskRunner()) {}
+  ~HashPrefixStoreWorker() = default;
+
+ private:
+  static scoped_refptr<base::SequencedTaskRunner> CreateTaskRunner() {
+    return base::ThreadPool::CreateSequencedTaskRunner(
+        {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+         base::TaskShutdownBehavior::BLOCK_SHUTDOWN});
```
Comment (atuchin-m) [inline ios/browser/api/brave_rewards/brave_rewards_api.mm:61]: maybe `SKIP_SHUTDOWN`?
Also does it make sense to reuse one of the current task runner to simplify things?

File: components/brave_rewards/core/engine/hash_prefix_store.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+class HashPrefixStore : public mojom::HashPrefixStore {
+ public:
+  explicit HashPrefixStore(base::FilePath path);
+  ~HashPrefixStore() override;
+
+  HashPrefixStore(const HashPrefixStore&) = delete;
+  HashPrefixStore& operator=(const HashPrefixStore&) = delete;
+
+  // Opens the hash prefix file, if not already open.
+  bool Open();
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_store.h:32]: Let's move internal methods to the private section.

File: components/brave_rewards/core/engine/hash_prefix_iterator.h (hunk 1 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_iterator.h:16]: This is a rename (and cleanup) of an existing class currently used by `PrefixListReader`.

File: components/brave_rewards/core/engine/hash_prefix_store.cc (hunk 2 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store.cc:55]: See `IsValidFile` for span-length validation.

File: components/brave_rewards/core/engine/hash_prefix_store.h (hunk 4 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store.h:32]: This method is intentionally public. It's used by the unit tests for instance, but in principle could be used by a client to manually control opening/closing.

File: components/brave_rewards/core/mojom/rewards_database.mojom [hunk 5]
```diff
@@ -71,3 +71,17 @@ struct DBCommandResponse {
...
+interface RewardsDatabase {
+  RunDBTransaction(DBTransaction transaction) => (DBCommandResponse response);
+};
+
+struct HashPrefixData {
+  string prefixes;
+  uint32 prefix_size;
+};
+
+interface HashPrefixStore {
```
Comment (atuchin-m) [inline components/brave_rewards/core/mojom/rewards_database.mojom:84]: Do we use actually use this interface for IPC? If not why just not use sequenceBound?

Comment (zenparsing) [reply components/brave_rewards/core/mojom/rewards_database.mojom:84]: I considered passing this interface to the Rewards engine, but decided to leave that for a future improvement. In the meantime, it is used to run the interface on a different thread (via `RemoteWorker`). However, if you aren't comfortable with that approach, I can use `SequenceBound` instead.

File: components/brave_rewards/content/rewards_service_impl.cc [hunk 6]
```diff
@@ -236,12 +237,14 @@ void DeferCallback(base::Location location, Callback callback, Args&&... args) {
 // read comment about file pathes at src\base\files\file_path.h
 #if BUILDFLAG(IS_WIN)
 const base::FilePath::StringType kDiagnosticLogPath(L"Rewards.log");
+const base::FilePath::StringType kCreatorPrefixStore(L"RewardsCreators.db");
```
Comment (emerick) [inline components/brave_rewards/content/rewards_service_impl.cc:240]: Can we keep these in alphabetical order?

**Review by emerick — APPROVED** [review]
++rewards

File: components/brave_rewards/content/rewards_service_impl.cc (hunk 6 above)
Comment (zenparsing) [reply components/brave_rewards/content/rewards_service_impl.cc:239]: Fixed.

File: components/brave_rewards/core/engine/hash_prefix_iterator.h (hunk 1 above)
Comment (atuchin-m) [reply components/brave_rewards/core/engine/hash_prefix_iterator.h:16]: O, thanks, I missed that part.
I see it uses to make sanity check that the first few prefixes are in order.  That code can be moved to `UpdatePrefixes`.
If we can rid off it completely, it will be great.

File: components/brave_rewards/core/mojom/rewards_database.mojom (hunk 5 above)
Comment (atuchin-m) [reply components/brave_rewards/core/mojom/rewards_database.mojom:84]: Ok, it's up to you.
From my perspective it adds extra complexity, but could be useful in some scenarios.

**Review by atuchin-m — COMMENTED** [review]
In general the PR looks good to me.
Added a few comments to address.

File: components/brave_rewards/core/engine/database/migration/migration_v41.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_REWARDS_CORE_ENGINE_DATABASE_MIGRATION_MIGRATION_V41_H_
+#define BRAVE_COMPONENTS_BRAVE_REWARDS_CORE_ENGINE_DATABASE_MIGRATION_MIGRATION_V41_H_
+
+namespace brave_rewards::internal::database::migration {
+
+// Migration 41 removes the publisher prefix list table after storage was moved
+// to a separate file.
+inline constexpr char v41[] = R"sql(
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/engine/database/migration/migration_v41.h:13]: We drop the table on Rewards engine startup, and also reset the update timer so that the file will be automatically downloaded very soon after startup. There may be a small amount of time during startup when there is no prefix list locally. This does not seem to affect the badge displaying on startup.

File: components/brave_rewards/core/remote_worker.h [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    requires(std::is_base_of_v<T, Impl>)
+  void BindRemote(Args&&... args) {
+    remote_.reset();
+    CreateRemoteWorker<Impl>(task_runner_, remote_.BindNewPipeAndPassReceiver(),
+                             std::forward<Args>(args)...);
+  }
+
+ private:
+  mojo::Remote<T> remote_;
+  scoped_refptr<base::SequencedTaskRunner> task_runner_;
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/remote_worker.h]: Fixed.

File: components/brave_rewards/core/remote_worker.h [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                                std::forward<Args>(args)...));
+}
+
+// A convenience wrapper around mojo::Remote that allows running a Mojo
+// interface implementation on a worker sequence.
+template <typename T>
+class RemoteWorker {
+ public:
+  explicit RemoteWorker(scoped_refptr<base::SequencedTaskRunner> task_runner)
+      : task_runner_(task_runner) {}
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/remote_worker.h]: Fixed.

File: components/brave_rewards/core/engine/hash_prefix_store.cc (hunk 2 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store.cc:73]: Good catch! Added a prefix count to the header. The file now looks like:

```
[version:u32] [prefix_size:u32] [prefix_count:u64] [data...]
```

File: components/brave_rewards/core/engine/hash_prefix_iterator.h [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  bool operator<=(const HashPrefixIterator& rhs) const {
+    return index_ <= rhs.index_;
+  }
+
+ private:
+  std::string_view PrefixAt(size_t index) const {
+    return data_.substr(index_ * size_, size_);
+  }
+
+  std::string_view data_;
```
[the remark this answers is not in the document]
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_iterator.h]: Fixed.

File: components/brave_rewards/core/engine/hash_prefix_store_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+class RewardsHashPrefixStoreTest : public testing::Test {
+ public:
+  void SetUp() override {
+    Test::SetUp();
+    ASSERT_TRUE(temp_dir_.CreateUniqueTempDir());
+    store_path_ = temp_dir_.GetPath().AppendASCII("prefixes.store");
+  }
+
+  void TearDown() override {
+    base::DeleteFile(store_path_);
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_store_unittest.cc:34]: Do we need to call this explicitly?
```
// An object representing a temporary / scratch directory that should be
// cleaned up (recursively) when this object goes out of scope. 
```
https://source.chromium.org/chromium/chromium/src/+/main:base/files/scoped_temp_dir.h;l=8-14

File: components/brave_rewards/core/engine/hash_prefix_store_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+TEST_F(RewardsHashPrefixStoreTest, ZeroPrefixes) {
+  auto header = MakeFileHeader(1, 4, 0);
+  base::WriteFile(store_path_, base::as_byte_span(header));
+  HashPrefixStore store(store_path_);
+  EXPECT_TRUE(store.Open());
+  EXPECT_FALSE(store.ContainsPrefix("test-value"));
+}
+
+TEST_F(RewardsHashPrefixStoreTest, WithPrefixes) {
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_store_unittest.cc:143]: I would also add a case to test updating a non-empty file, to check the sequence `Open() => ContainsPrefix() => Close() => UpdatePrefixes() => ContainsPrefix()`

File: components/brave_rewards/core/engine/hash_prefix_store.h [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+  // mojom::HashPrefixStore:
+  void UpdatePrefixes(mojom::HashPrefixDataPtr prefix_data,
+                      UpdatePrefixesCallback callback) override;
+
+  void ContainsPrefix(const std::string& value,
+                      ContainsPrefixCallback callback) override;
+
+ private:
+  base::FilePath file_path_;
```
Comment (atuchin-m) [inline components/brave_rewards/core/engine/hash_prefix_store.h:52]: nit: add const

**Review by atuchin-m — APPROVED** [review]
LGTM with a few last nits.
Thanks!

File: components/brave_rewards/core/engine/hash_prefix_store.h (hunk 13 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store.h]: Fixed.

File: components/brave_rewards/core/engine/hash_prefix_store_unittest.cc (hunk 12 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store_unittest.cc:138]: Added a test case.

File: components/brave_rewards/core/engine/hash_prefix_store_unittest.cc (hunk 11 above)
Comment (zenparsing) [reply components/brave_rewards/core/engine/hash_prefix_store_unittest.cc]: Removed.

File: components/brave_rewards/core/mojom/rewards_database.mojom (hunk 5 above)
Comment (zenparsing) [reply components/brave_rewards/core/mojom/rewards_database.mojom]: I decided to go with `SequenceBound` - thanks for the suggestion.

File: ios/browser/api/brave_rewards/brave_rewards_api.mm [hunk 14]
```diff
@@ -172,6 +176,10 @@ - (instancetype)initWithStateStoragePath:(NSString*)path {
         base::SequenceBound<brave_rewards::internal::RewardsDatabase>(
             databaseQueue, base::FilePath(dbPath));
 
+    creatorPrefixStore = base::SequenceBound<HashPrefixStore>(
+        databaseQueue,
+        base::FilePath([self creatorPrefixStorePath].UTF8String));
```
Comment (kylehickinson) [inline ios/browser/api/brave_rewards/brave_rewards_api.mm:181]: minor nits:
- we should add an underscore prefix to the ivar (e.g. `_creatorPrefixStore`)
- we should use `base::SysNSStringToUTF8` instead of using `UTF8String` directly

**Review by kylehickinson — APPROVED** [review]

File: ios/browser/api/brave_rewards/brave_rewards_api.mm (hunk 14 above)
Comment (zenparsing) [reply ios/browser/api/brave_rewards/brave_rewards_api.mm]: Fixed - thanks!
